### PR TITLE
Improvement: restart sbt server

### PR DESCRIPTION
**Pull request by kasiaMarek**
Add restart build sever for `sbt` build server. Connected to: https://github.com/scalameta/metals/issues/5313

### CODE REVIEW

File: tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala [hunk 1]
```diff
@@ -32,10 +33,13 @@ class SbtServerSuite
   val scalaVersion = V.scala213
   val buildTool: SbtBuildTool = SbtBuildTool(None, () => userConfig)
   var compilationCount = 0
+  val bspStartLogs: ListBuffer[String] = ListBuffer.empty[String]
+  override val handleBspStartupErrorMessage = bspStartLogs.append
```
Comment (tgodzik) [inline tests/slow/src/test/scala/tests/sbt/SbtServerSuite.scala:37]: Can you use `workspaceLogMessages` instead? I would avoid passing around the function if possible

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala [hunk 2]
```diff
@@ -1579,9 +1581,51 @@ class MetalsLspService(
...
   }
 
-  def shutDownBloop(): Boolean = bloopServers.shutdownServer()
-
-  def isBloop(): Boolean = bspSession.exists(_.main.isBloop)
+  /**
+   * We only support restarting bloop and sbt servers.
+   */
+  def restartBspServer(): Future[Boolean] = {
+    def emitMessage(msg: String, msgType: MessageType = MessageType.Info) = {
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:1588]: Doesn't `scribe.info` do the same?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala [hunk 3]
```diff
@@ -1579,9 +1581,51 @@ class MetalsLspService(
...
+          _ <- autoConnectToBuildServer()
+        } yield handleRestartResult(res, "bloop")
+      case Some(session) if session.main.isSbt =>
+        for {
+          currentBuildTool <- supportedBuildTool
+          res <- currentBuildTool match {
+            case Some(sbt: SbtBuildTool) =>
+              for {
+                _ <- disconnectOldBuildServer()
+                code <- sbt.shutdownSbtServer(shellRunner, workspace = folder)
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:1613]: Could we this for each server and return false if it was not possible?

Add `def shutdown() = false` method to BuildTool and override in SbtBuildTool and maybe MillBuildTool? `mill shutdown` should work there

**Review by tgodzik — CHANGES_REQUESTED** [review]
Thanks for the great work!

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala (hunk 2 above)
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:1588]: I'm not logging. I'm showing the info to the user.

**Comment by tgodzik** [comment]
We also have `build/shutdown` method, which might be used for all the other BSP server aside from Bloop maybe?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala (hunk 2 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:1588]: Shouldn't it be `showMessage` ?

**Comment by kasiaMarek** [comment]
> We also have `build/shutdown` method, which might be used for all the other BSP server aside from Bloop maybe?

Yeah, but it doesn't really shutdown the sever. At least not for `sbt` (or `bloop`). Btw we call it every time we reconnect. So I guess it should be enough to reconnect for all other build servers (that is what we did before).

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala (hunk 2 above)
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:1588]: Yeah...

**Comment by tgodzik** [comment]
> > We also have `build/shutdown` method, which might be used for all the other BSP server aside from Bloop maybe?
> 
> Yeah, but it doesn't really shutdown the sever. At least not for `sbt` (or `bloop`). Btw we call it every time we reconnect. So I guess it should be enough to reconnect for all other build servers (that is what we did before).

I think reconnecting as default should be fine.

**Review by tgodzik — APPROVED** [review]
LGTM!

**Pull request merged**